frontend/model/contracts/chatroom.js

```diff
@@ -327,36 +343,88 @@ sbp('chelonia/defineContract', {
       }
     },
-    'gi.contracts/chatroom/deleteMessage': {
-      validate: objectMaybeOf({
-
-      }),
-      process ({ data, meta }, { state }) {
-
-      }
-    },
     'gi.contracts/chatroom/editMessage': {
-      validate: objectMaybeOf({
-
-      }),
+      validate: (data, { state, meta }) => {
+        objectOf({
+          id: string,
+          text: string
+        })(data)
+        // TODO: need to check if the meta.username === message.from
+      },
       process ({ data, meta }, { state }) {
-
+        const msgIndex = findMessageIdx(data.id, state.messages)
+        if (msgIndex >= 0) {
+          state.messages[msgIndex].text = data.text
+          state.messages[msgIndex].updatedDate = meta.createdDate
+        }
+      },
+      sideEffect ({ contractID, data }, { state }) {
+        emitMessageEvent({
+          type: MESSAGE_ACTION_TYPES.EDIT_MESSAGE,
+          contractID,
+          hash: data.id,
+          state
+        })
       }
     },
-    'gi.contracts/chatroom/addEmoticon': {
-      validate: objectMaybeOf({
-
+    'gi.contracts/chatroom/deleteMessage': {
+      validate: objectOf({
+        id: string
       }),
       process ({ data, meta }, { state }) {
-
+        const msgIndex = findMessageIdx(data.id, state.messages)
+        if (msgIndex >= 0) {
+          state.messages.splice(msgIndex, 1)
+        }
+      },
+      sideEffect ({ contractID, data }, { state }) {
+        emitMessageEvent({
+          type: MESSAGE_ACTION_TYPES.DELETE_MESSAGE,
+          contractID,
+          hash: data.id,
+          state
+        })
       }
     },
-    'gi.contracts/chatroom/deleteEmoticon': {
-      validate: objectMaybeOf({
-
+    'gi.contracts/chatroom/makeEmotion': {
+      validate: objectOf({
+        id: string,
+        emoticon: string
       }),
-      process ({ data, meta }, { state }) {
-
+      process ({ data, meta, contractID }, { state }) {
+        const { id, emoticon } = data
+        const msgIndex = findMessageIdx(id, state.messages)
+        if (msgIndex >= 0) {
+          let emoticons = state.messages[msgIndex].emoticons || {}
+          if (emoticons[emoticon]) {
+            const alreadyAdded = emoticons[emoticon].indexOf(meta.username)
+            if (alreadyAdded >= 0) {
+              emoticons[emoticon].splice(alreadyAdded, 1)
+              if (!emoticons[emoticon].length) {
+                delete emoticons[emoticon]
+                if (!Object.keys(emoticons).length) {
+                  emoticons = null
+                }
+              }
+            } else {
+              emoticons[emoticon].push(meta.username)
+            }
+          } else {
+            emoticons[emoticon] = [meta.username]
+          }
+          if (emoticons) {
+            Vue.set(state.messages[msgIndex], 'emoticons', emoticons)
+          } else {
+            Vue.delete(state.messages[msgIndex], 'emoticons')
+          }
```

Comment:
This code could be simplified & shortened by leaving an empty list when there are no more emoticons left. I don't think it needs the code that completely deletes the entire emoticons list.

Reply:
Of course, if there are `emoticons` empty list, there won't be any thing wrong.
All things will work well and nothing breaks. But no need to save redundant thing, don't we?
I think the problem is to shorten the code lines instead of leaving empty array.

Reply:
Well, shorter code is easier to read, understand, and has fewer bugs...

What is the harm with leaving an empty array? These messages will get cleared out of memory eventually anyway.